### PR TITLE
feat: [IOCOM-253,IOCOM-259] Refactor `lollipopFetch` to extract and export the signature creation

### PR DESCRIPTION
**Pull request by Vangaorth**
## Short description
This PR does an extract-method refactoring on the lollipopFetch body, exporting a new callable function, lollipopRequestInit, that produces the signed lollipop headers.

## List of changes proposed in this pull request
- `ts/features/lollipop/utils/fetch.ts`:
  - `lollipopFetch`: 
    - this function's body has been extracted into the `lollipopRequestInit` function
  - `lollipopRequestInit`: 
    - contains the old `lollipopFetch`'s body
    - it has been refactored in order to throw an exception if something goes amiss

## How to test
Using the io-dev-api-server, check that the lollipop playground is working properly

### CODE REVIEW

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    expect(init).toStrictEqual({
+      ...initToMatch,
+      body: "body",
+      headers: {
+        "Content-Digest": testContentDigest,
+        ...initToMatch.headers
+      }
+    });
+  });
+  it("should throw if no keyTag is set", async () => {
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:90]: We could add a space to be consistent with the whole file? What do you think? :)

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    try {
+      return await lollipopRequestInit(
+        lollipopConfig,
+        { ...keyInfo, keyTag: undefined },
+        fullUrl,
+        requestInit
+      );
+    } catch (e) {
+      return expect(`${e}`).toMatch(configurationError);
+    }
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:100]: Why not using `.rejects.toEqual` like did in the last test?

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    try {
+      return await lollipopRequestInit(
+        lollipopConfig,
+        { ...keyInfo, publicKey: undefined },
+        fullUrl,
+        requestInit
+      );
+    } catch (e) {
+      return expect(`${e}`).toMatch(configurationError);
+    }
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:113]: ditto

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    try {
+      return await lollipopRequestInit(
+        lollipopConfig,
+        keyInfo,
+        fullUrlRequest,
+        requestInit
+      );
+    } catch (e) {
+      return expect(`${e}`).toMatch(configurationError);
+    }
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:127]: ditto

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+  it("should throw if no headers are set", async () => {
+    try {
+      return await lollipopRequestInit(lollipopConfig, keyInfo, fullUrl, {
+        ...requestInit,
+        headers: undefined
+      });
+    } catch (e) {
+      return expect(`${e}`).toMatch(configurationError);
+    }
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:138]: ditto

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    const init = await lollipopRequestInit(
+      lollipopConfig,
+      keyInfo,
+      fullUrl,
+      requestInit
+    );
+    expect(init).toStrictEqual(testInit(getUnixTimestamp()));
+  });
+
+  it("should succeed with also body signature", async () => {
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:67]: Testing all the features is good! We could add a test for the `customContentToSign` too. What do you think?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]
Just little changes and maybe one more test.

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts (hunk 1 above)
Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:119]: Definitely

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts (hunk 2 above)
Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts]: My bad, I misunderstood the original recommendation. Corrected

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts (hunk 3 above)
Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts]: Done

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts (hunk 4 above)
Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts]: Done

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts (hunk 6 above)
Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts]: Good idea. Done it, check it out.

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    "signature-input": `sig1=(${
+      signBody ? '"content-digest" ' : ""
+    }"x-pagopa-lollipop-original-method" "x-pagopa-lollipop-original-url");created=${timestamp};nonce="nonce";alg="ecdsa-p256-sha256";keyid="SXn6l6BNlwAb60cJUKpvKB3H-UQbe2slQ_8LBR70cfA"`
+  },
+  method: "GET"
+});
+
+const testContentDigest =
+  "sha-256=:Iw2DWNyOiJC0xY3utikS7i8gNXrpKlzIYbmOaP4xrLU=:";
+
```
Comment (shadowsheep1) [inline ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:55]:
```suggestion

MockDate.set("2023-01-01T01:00:00");
const mockTimestamp = 1672534800;

describe("Check test mocks", () => {
  describe("timestamp from getUnixTimestamp()", () => {
    it(`should be ${mockTimestamp}`, () => {
      expect(getUnixTimestamp()).toBe(mockTimestamp);
    });
  });
});

```

To be more "time agnostic" against the `getUnixTimestamp` we could mock the test date. What do you think_

**Review by shadowsheep1 — APPROVED** [review]
Apart my last comment... LGTM! 💯

File: ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts (hunk 7 above)
Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:55]: It is a useful test but, IMHO, it does not belong here. Where would it be a good place to ~~place~~ put it?

Comment (shadowsheep1) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:55]: It's more a sanity check for the mock setup. Anyway for the "time agnostic" part, only line 56 it's enough.

Comment (Vangaorth) [reply ts/features/lollipop/utils/__tests__/lollipopRequestInit.test.ts:59]: Done it!

**Review by shadowsheep1 — APPROVED** [review]
LGTM

**Pull request merged**